From e43a7783c93a4f1be2695c5f1189d30f1a2faa80 Mon Sep 17 00:00:00 2001
From: Sergey Fedorov <barracuda@macos-powerpc.org>
Date: Mon, 27 Jul 2026 20:34:48 +0000
Subject: [PATCH 11/16] Implement wp_single_pixel_buffer_v1

Vendor the staging single-pixel-buffer-v1 protocol. OwlSinglePixelBuffer
is a 1x1 OwlBuffer subclass that fills its rect with the requested
premultiplied RGBA color; mpv uses this for solid backgrounds.
---
 Sources/Compositor/OwlSinglePixelBuffer.h     |  34 ++++++
 Sources/Compositor/OwlSinglePixelBuffer.m     |  76 +++++++++++++
 .../OwlWpSinglePixelBufferManagerV1.h         |  32 ++++++
 .../OwlWpSinglePixelBufferManagerV1.m         | 104 ++++++++++++++++++
 Sources/Protocol/single-pixel-buffer-v1.xml   |  76 +++++++++++++
 Sources/Server/OwlServer.m                    |   2 +
 6 files changed, 324 insertions(+)
 create mode 100644 Sources/Compositor/OwlSinglePixelBuffer.h
 create mode 100644 Sources/Compositor/OwlSinglePixelBuffer.m
 create mode 100644 Sources/Compositor/OwlWpSinglePixelBufferManagerV1.h
 create mode 100644 Sources/Compositor/OwlWpSinglePixelBufferManagerV1.m
 create mode 100644 Sources/Protocol/single-pixel-buffer-v1.xml

diff --git a/Sources/Compositor/OwlSinglePixelBuffer.h b/Sources/Compositor/OwlSinglePixelBuffer.h
new file mode 100644
index 0000000..e148c0a
--- /dev/null
+++ b/Sources/Compositor/OwlSinglePixelBuffer.h
@@ -0,0 +1,34 @@
+/* This file is part of Owl.
+ *
+ * Copyright © 2019-2021 Sergey Bugaev <bugaevc@gmail.com>
+ *
+ * Owl is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * Owl is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Owl.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#import "OwlBuffer.h"
+#import <Cocoa/Cocoa.h>
+#import <wayland-server.h>
+
+
+@interface OwlSinglePixelBuffer : OwlBuffer {
+    CGFloat _red, _green, _blue, _alpha;
+}
+
+- (id) initWithResource: (struct wl_resource *) resource
+                       r: (uint32_t) r
+                       g: (uint32_t) g
+                       b: (uint32_t) b
+                       a: (uint32_t) a;
+
+@end
diff --git a/Sources/Compositor/OwlSinglePixelBuffer.m b/Sources/Compositor/OwlSinglePixelBuffer.m
new file mode 100644
index 0000000..7661cc7
--- /dev/null
+++ b/Sources/Compositor/OwlSinglePixelBuffer.m
@@ -0,0 +1,76 @@
+/* This file is part of Owl.
+ *
+ * Copyright © 2019-2021 Sergey Bugaev <bugaevc@gmail.com>
+ *
+ * Owl is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * Owl is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Owl.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#import "OwlSinglePixelBuffer.h"
+
+
+@implementation OwlSinglePixelBuffer
+
+- (id) initWithResource: (struct wl_resource *) resource
+                       r: (uint32_t) r
+                       g: (uint32_t) g
+                       b: (uint32_t) b
+                       a: (uint32_t) a
+{
+    self = [super initWithResource: resource];
+
+    // Per the protocol, these values use premultiplied alpha and are
+    // expressed as a fraction of UINT32_MAX.
+    _alpha = a / 4294967295.0;
+    _red = r / 4294967295.0;
+    _green = g / 4294967295.0;
+    _blue = b / 4294967295.0;
+
+    return self;
+}
+
+- (NSSize) size {
+    return NSMakeSize(1, 1);
+}
+
+- (BOOL) needsGLForRendering {
+    return NO;
+}
+
+- (NSColor *) color {
+    // Un-premultiply so NSColor gets straight alpha components.
+    if (_alpha == 0.0) {
+        return [NSColor colorWithCalibratedRed: 0.0 green: 0.0 blue: 0.0 alpha: 0.0];
+    }
+    return [NSColor colorWithCalibratedRed: _red / _alpha
+                                      green: _green / _alpha
+                                       blue: _blue / _alpha
+                                      alpha: _alpha];
+}
+
+- (void) drawInRect: (NSRect) rect {
+    [[self color] set];
+    NSRectFillUsingOperation(rect, NSCompositeSourceOver);
+}
+
+- (NSImage *) createNSImage {
+    NSSize size = NSMakeSize(1, 1);
+    NSImage *image = [[NSImage alloc] initWithSize: size];
+    [image lockFocus];
+    [[self color] set];
+    NSRectFillUsingOperation(NSMakeRect(0, 0, 1, 1), NSCompositeSourceOver);
+    [image unlockFocus];
+    return image;
+}
+
+@end
diff --git a/Sources/Compositor/OwlWpSinglePixelBufferManagerV1.h b/Sources/Compositor/OwlWpSinglePixelBufferManagerV1.h
new file mode 100644
index 0000000..83bc8d9
--- /dev/null
+++ b/Sources/Compositor/OwlWpSinglePixelBufferManagerV1.h
@@ -0,0 +1,32 @@
+/* This file is part of Owl.
+ *
+ * Copyright © 2019-2021 Sergey Bugaev <bugaevc@gmail.com>
+ *
+ * Owl is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * Owl is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Owl.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#import "OwlGlobal.h"
+#import <Cocoa/Cocoa.h>
+#import <wayland-server.h>
+
+
+@interface OwlWpSinglePixelBufferManagerV1 : NSObject <OwlGlobal> {
+    struct wl_resource *_resource;
+}
+
+- (id) initWithResource: (struct wl_resource *) resource;
+
++ (void) addGlobalToDisplay: (struct wl_display *) display;
+
+@end
diff --git a/Sources/Compositor/OwlWpSinglePixelBufferManagerV1.m b/Sources/Compositor/OwlWpSinglePixelBufferManagerV1.m
new file mode 100644
index 0000000..087ad59
--- /dev/null
+++ b/Sources/Compositor/OwlWpSinglePixelBufferManagerV1.m
@@ -0,0 +1,104 @@
+/* This file is part of Owl.
+ *
+ * Copyright © 2019-2021 Sergey Bugaev <bugaevc@gmail.com>
+ *
+ * Owl is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * Owl is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Owl.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#import "OwlWpSinglePixelBufferManagerV1.h"
+#import "OwlSinglePixelBuffer.h"
+#import "single-pixel-buffer-v1.h"
+#import <wayland-server.h>
+
+
+@implementation OwlWpSinglePixelBufferManagerV1
+
+static void single_pixel_buffer_manager_destroy_handler(
+    struct wl_client *client,
+    struct wl_resource *resource
+) {
+    wl_resource_destroy(resource);
+}
+
+static void single_pixel_buffer_manager_create_u32_rgba_buffer_handler(
+    struct wl_client *client,
+    struct wl_resource *resource,
+    uint32_t id,
+    uint32_t r,
+    uint32_t g,
+    uint32_t b,
+    uint32_t a
+) {
+    struct wl_resource *buffer_resource = wl_resource_create(
+        client,
+        &wl_buffer_interface,
+        1,
+        id
+    );
+    OwlSinglePixelBuffer *buffer = [OwlSinglePixelBuffer alloc];
+    [[buffer initWithResource: buffer_resource
+                             r: r
+                             g: g
+                             b: b
+                             a: a] release];
+}
+
+static const struct wp_single_pixel_buffer_manager_v1_interface
+single_pixel_buffer_manager_impl = {
+    .destroy = single_pixel_buffer_manager_destroy_handler,
+    .create_u32_rgba_buffer = single_pixel_buffer_manager_create_u32_rgba_buffer_handler
+};
+
+static void single_pixel_buffer_manager_destroy(struct wl_resource *resource) {
+    OwlWpSinglePixelBufferManagerV1 *self = wl_resource_get_user_data(resource);
+    [self release];
+}
+
+- (id) initWithResource: (struct wl_resource *) resource {
+    _resource = resource;
+    wl_resource_set_implementation(
+        resource,
+        &single_pixel_buffer_manager_impl,
+        [self retain],
+        single_pixel_buffer_manager_destroy
+    );
+    return self;
+}
+
+static void single_pixel_buffer_manager_bind(
+    struct wl_client *client,
+    void *data,
+    uint32_t version,
+    uint32_t id
+) {
+    struct wl_resource *resource = wl_resource_create(
+        client,
+        &wp_single_pixel_buffer_manager_v1_interface,
+        version,
+        id
+    );
+    [[[OwlWpSinglePixelBufferManagerV1 alloc] initWithResource: resource] release];
+}
+
++ (void) addGlobalToDisplay: (struct wl_display *) display {
+    wl_global_create(
+        display,
+        &wp_single_pixel_buffer_manager_v1_interface,
+        1,
+        NULL,
+        single_pixel_buffer_manager_bind
+    );
+}
+
+@end
diff --git a/Sources/Protocol/single-pixel-buffer-v1.xml b/Sources/Protocol/single-pixel-buffer-v1.xml
new file mode 100644
index 0000000..bcbd038
--- /dev/null
+++ b/Sources/Protocol/single-pixel-buffer-v1.xml
@@ -0,0 +1,76 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<protocol name="single_pixel_buffer_v1">
+  <copyright>
+    Copyright © 2022 Simon Ser
+
+    Permission is hereby granted, free of charge, to any person obtaining a
+    copy of this software and associated documentation files (the "Software"),
+    to deal in the Software without restriction, including without limitation
+    the rights to use, copy, modify, merge, publish, distribute, sublicense,
+    and/or sell copies of the Software, and to permit persons to whom the
+    Software is furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice (including the next
+    paragraph) shall be included in all copies or substantial portions of the
+    Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+    DEALINGS IN THE SOFTWARE.
+  </copyright>
+
+  <description summary="single pixel buffer factory">
+    This protocol extension allows clients to create single-pixel buffers.
+
+    Compositors supporting this protocol extension should also support the
+    viewporter protocol extension. Clients may use viewporter to scale a
+    single-pixel buffer to a desired size.
+
+    Warning! The protocol described in this file is currently in the testing
+    phase. Backward compatible changes may be added together with the
+    corresponding interface version bump. Backward incompatible changes can
+    only be done by creating a new major version of the extension.
+  </description>
+
+  <interface name="wp_single_pixel_buffer_manager_v1" version="1">
+    <description summary="global factory for single-pixel buffers">
+      The wp_single_pixel_buffer_manager_v1 interface is a factory for
+      single-pixel buffers.
+    </description>
+
+    <request name="destroy" type="destructor">
+      <description summary="destroy the manager">
+        Destroy the wp_single_pixel_buffer_manager_v1 object.
+
+        The child objects created via this interface are unaffected.
+      </description>
+    </request>
+
+    <request name="create_u32_rgba_buffer">
+      <description summary="create a 1×1 buffer from 32-bit RGBA values">
+        Create a single-pixel buffer from four 32-bit RGBA values.
+
+        Unless specified in another protocol extension, the RGBA values use
+        pre-multiplied alpha.
+
+        The width and height of the buffer are 1.
+
+        The r, g, b and a arguments valid range is from UINT32_MIN (0)
+        to UINT32_MAX (0xffffffff).
+        
+        These arguments should be interpreted as a percentage, i.e.
+        - UINT32_MIN = 0% of the given color component
+        - UINT32_MAX = 100% of the given color component
+      </description>
+      <arg name="id" type="new_id" interface="wl_buffer"/>
+      <arg name="r" type="uint" summary="value of the buffer's red channel"/>
+      <arg name="g" type="uint" summary="value of the buffer's green channel"/>
+      <arg name="b" type="uint" summary="value of the buffer's blue channel"/>
+      <arg name="a" type="uint" summary="value of the buffer's alpha channel"/>
+    </request>
+  </interface>
+</protocol>
diff --git a/Sources/Server/OwlServer.m b/Sources/Server/OwlServer.m
index 6cab3b6..4749182 100644
--- a/Sources/Server/OwlServer.m
+++ b/Sources/Server/OwlServer.m
@@ -26,6 +26,7 @@
 #import "OwlXdgWmBase.h"
 #import "OwlZxdgDecorationManagerV1.h"
 #import "OwlWpContentTypeManagerV1.h"
+#import "OwlWpSinglePixelBufferManagerV1.h"
 #import "OwlSeat.h"
 #import "OwlOutput.h"
 #import "OwlWlDataDeviceManager.h"
@@ -129,6 +130,7 @@
     [OwlXdgWmBase addGlobalToDisplay: _display];
     [OwlZxdgDecorationManagerV1 addGlobalToDisplay: _display];
     [OwlWpContentTypeManagerV1 addGlobalToDisplay: _display];
+    [OwlWpSinglePixelBufferManagerV1 addGlobalToDisplay: _display];
     [OwlWlDataDeviceManager addGlobalToDisplay: _display];
     [OwlSeat addGlobalToDisplay: _display];
     [OwlOutput addGlobalToDisplay: _display];
-- 
2.43.0

